gemini/sample-apps/llamaindex-rag/src/app/main.py

```diff
@@ -0,0 +1,484 @@
+        logger.info(databases)
+        if active_db_name is None:
+            databases.insert(0, None)
+        else:
+            databases.remove(active_db_name)
+            databases.insert(0, active_db_name)
+        return databases
+
+    except Exception as e:
+        logger.info(f"An error occurred: {e}")
+        return []
+
+
+@app.post("/list_firestore_collections")
+async def list_firestore_collections(
+    db_name: dict, index_manager=Depends(get_index_manager)
+):
+    """
+    List all collections in a Firestore database.
+
+    :param project_id: Your Google Cloud project ID
+    :param database_id: The database ID. Use "(default)" for the default database.
+    :return: A list of collection IDs
+    """
+
+    def get_prefixes(string_list):
+        suffixes = ["_metadata", "_data", "_ref_doc_info"]
+        prefixes = []
+
+        for s in string_list:
+            for suffix in suffixes:
+                if s.endswith(suffix):
+                    prefixes.append(s[: -len(suffix)])
+                    break
+            else:
+                prefixes.append(s)
+
+        return prefixes
+
+    # Initialize Firestore client
+    if db_name["firestore_db_name"]:
+        db = firestore.Client(
+            project=index_manager.project_id, database=db_name["firestore_db_name"]
+        )
+    else:
+        return []
+    try:
+        # Get all collections
+        collections = db.collections()
+        collection_info = [collection.id for collection in collections]
+        collection_info = list(set(get_prefixes(collection_info)))
+        active_firestore_namespace = index_manager.firestore_namespace
+        if active_firestore_namespace is None:
+            collection_info.insert(0, None)
+        else:
+            collection_info.remove(active_firestore_namespace)
+            collection_info.insert(0, active_firestore_namespace)
+        return collection_info
+
+    except PermissionDenied:
+        print(
+            f"Permission denied. Make sure you have the necessary permissions to access Firestore in project {PROJECT_ID}"
+        )
+        return []
+    except Exception as e:
+        print(f"An error occurred: {e}")
+        return []
+
+
+@app.get("/get_current_index_info")
+async def get_current_index_info(index_manager=Depends(get_index_manager)):
+    return index_manager.get_current_index_info()
+
+
+@app.post("/update_index")
+async def update_index(
+    index_update: IndexUpdate, index_manager=Depends(get_index_manager)
+):
+    try:
+        index_manager.set_current_indexes(
+            index_update.base_index_name,
+            index_update.base_endpoint_name,
+            index_update.qa_index_name,
+            index_update.qa_endpoint_name,
+            index_update.firestore_db_name,
+            index_update.firestore_namespace,
+        )
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e))
+
+
+@app.post("/query_rag")
+async def query_rag(
```

Comment:
Try to simplify this as possible

Reply:
This route is where the core RAG logic is. Unfortunately since we are exposing an experimental interface, this necessitates having complex if/then logic to support the various configurations the user decides on.